go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,102 @@
+package libovsdbops
+
+import (
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+const (
+	// db ownership ExternalIDs
+	// when possible, use DbIndex and function from this file instead of directly using the following ExternalIDs
+	OwnerController = "owner_controller"
+	OwnerObjectType = "owner_object_type"
+	OwnerNamespace  = "owner_namespace"
+	OwnerName       = "owner_name"
+	InternalID      = "internal_id"
```

Comment:
Thanks for the thorough review! That is exactly what I wanted to discuss :)
1. domain prefix - we can surely add that, but I wasn't sure if that prefix has any real context, since the db itself together with all controllers is a part of ovn-kubernetes code, so I guess all db entries are considered to be `k8s.ovn.org`. Maybe you envision some examples of another prefixes that can be used in the future different from it?
2. service controller compatibility - existing ids are not enough, as you mentioned we do need separate namespace and name, also `/kind`, which is filled with `obj.GetObjectKind().GroupVersionKind().GroupKind()` is not enough, since some db objects are not directly owned by k8s objects, e.g. `EgressFirewallDNSOwnerType` is the owner for dns entry, that can be shared by multiple db objects, so we can either use existing `/kind` id, but change its semantics (not only use real api server kinds), or update the id to not refer to the kind specifically.
So it looks like some changes for service controller owned objects will need to be done anyway, therefore we can figure out how big those changes will be, when we finalize the ids we need.
3. use external_ids as db indexes - new `/id` primary index sounds good, especially if it will improve performance
4. Using hyphens - definitely will do
5. Many controllers updating the same row - I really didn't consider this to be an option. For address sets, when something like this happens it is more of a bug than a feature, but I guess there are some better examples. Maybe if you can share some of them, it will help me to update the docs in a more specific way.
6. Additional external ids - controllers can add and use additional ids even now, they only need to make sure that existing set of externalIDs is unique, since it may be used as an index. Actually addressSetFactory adds its own `ipFamily` external id, that is nor overridden by dbIndex (or at least I didn't mean to :D) So I don't understand the example probably, can you explain a bit more what is the problem here?
7. I don't have enough knowledge of how new secondary controllers will look like, so what kind of `controller-type` do you expect to see and what kind of objects do you think it can own/modify?

Reply:
1. I think it's good practice and future proof and we already have a sample from the service controller of someone thinking along those lines. Also it helps noticing who is defining an external id. There are some already defined by OVN (ic-learned-route, neutron:network_name, neutron:port_name, etc...). Also note there are circumstances where northd copies external ids to the SBDB actually increasing the scope of the external ids further of what you might have initially thought.
2. OK. We can align where it makes sense now then, maybe with `controller` and removing the `owner_` prefix?
5. On things like switches (references to ports, ACLs, LBs) and routers (references to LBs, NATs, policies, routes) it might be more likely to happen. 
6. Well, that's even true before this PR. But what we see in this PR is some kind of framework/utility to help controllers handle the index of rows they work with. So exactly when is this DbIndex type ought to be used? And when it's not ought to be used? This is confusing, but if this DbIndex is thought out to define a common set of indexes, and rows can be shared by controllers in a way that we are interested to have the same set of indexes but with different values for different controllers, then we might want to future-proof for this as well even if today is not something that we will use.
7.  Pretty much the same as the default network controller. For example to create the address set factory you use
```
const DefaultNetworkControllerName = "default-network-controller"
```
which ends up being the controller name in the external id.
There is only one `default-network-controller` because there is only ever one default network. But we will have secondary networks, they can be of different types: "L2", "L3", "L3 with uplink". For each type, we can have multiple networks with a different name. For each network we will have a separate controller. These controllers will listen to the same events as the default network controller and create separate OVN things as a result, like other address sets, logical ports, etc.
So "default-network-controller" kinda combines the network name and type because there is only ever one primary cluster wide network. But for a secondary network we can have a controller of type "secondary-l3-network-controller". And we can have multiple controllers of that type with different names. Let's say we have a "storage" network and a "public" network both of type "secondary-l3-network-controller".

Reply:
I think any conflict can be avoided by just filtering out only rows with given `owner_controller` set, then controllers may even have the same additional ids (let's say some new controller uses `ipFamily` id same as addressSetFactory, but for some other purpose), then by just filtering out only rows with the correct `owner_controller`, controllers may be sure that the object is owned by them, and they know the semantics of this id.
So in general the whole id can be seen as a narrowing context like `<controller>.<owner>.<namespace>.<name>.<internal id>` (reverse this to get dns-name-like format :) )

Reply:
I might have oversimplified. I guess I am still talking from the point of view of controllers sharing the same row.